layer/profiles.cpp

```diff
@@ -3688,6 +3917,153 @@ bool JsonLoader::GetFormat(const Json::Value &formats, const std::string &format
     return valid;
 }
 
+bool QueueFamilyMatch(const VkQueueFamilyProperties &device_qfp, const VkQueueFamilyProperties &profile_qfp) {
+    if ((device_qfp.queueFlags & profile_qfp.queueFlags) != profile_qfp.queueFlags) {
+        return false;
+    }
+    if (device_qfp.queueCount < profile_qfp.queueCount) {
+        return false;
+    }
+    if (device_qfp.timestampValidBits != profile_qfp.timestampValidBits) {
+        return false;
+    }
+    if (device_qfp.minImageTransferGranularity.width >= profile_qfp.minImageTransferGranularity.width ||
+        device_qfp.minImageTransferGranularity.height >= profile_qfp.minImageTransferGranularity.height ||
+        device_qfp.minImageTransferGranularity.depth >= profile_qfp.minImageTransferGranularity.depth) {
+        return false;
+    }
+    return true;
+}
+
+bool JsonLoader::GetQueueFamilyProperties(const Json::Value &qf_props, ArrayOfVkQueueFamilyProperties *dest,
+                                          VkQueueFamilyGlobalPriorityPropertiesKHR *gp_dest,
+                                          VkVideoQueueFamilyProperties2KHR *vqfp_dest,
+                                          VkQueueFamilyCheckpointPropertiesNV *qfcp_dest,
+                                          VkQueueFamilyCheckpointProperties2NV *qfcp2_dest,
+                                          VkQueueFamilyQueryResultStatusProperties2KHR *qfqrsp_dest) {
+    for (const auto &name : qf_props.getMemberNames()) {
+        const auto &props = qf_props[name];
+        if (name == "VkQueueFamilyProperties") {
+            VkQueueFamilyProperties2 qfp = {};
+            qfp.sType = VK_STRUCTURE_TYPE_QUEUE_FAMILY_PROPERTIES_2;
+            for (const auto &feature : props["queueFlags"]) {
+                qfp.queueFamilyProperties.queueFlags |= StringToVkQueueFlags(feature.asString());
+            }
+            qfp.queueFamilyProperties.queueCount = props["queueCount"].asInt();
+            qfp.queueFamilyProperties.timestampValidBits = props["timestampValidBits"].asUInt();
+            const auto &minImagetransferGranularity = props["minImageTransferGranularity"];
+            qfp.queueFamilyProperties.minImageTransferGranularity.width = minImagetransferGranularity["width"].asUInt();
+            qfp.queueFamilyProperties.minImageTransferGranularity.height = minImagetransferGranularity["height"].asUInt();
+            qfp.queueFamilyProperties.minImageTransferGranularity.depth = minImagetransferGranularity["depth"].asUInt();
```

Comment:
What happens when the profile doesn't define a particular field of the structure?
Not sure what's the intention in the end. Should we just return zeros for those? Does the above code actually do that if the corresponding member isn't actually present in the JSON data, or will to raise an exception?
This actually leads to the wider question of what the default values of members in structs should be when the profile doesn't define a value. If zeros make sense always, then that's okay, but I'm not quite sure.

Reply:
Perhaps the values should then be taken from the compatible queue that will be used?

Reply:
That's what I'm not sure about, as in that case the layer would actually expose more capabilities than what's defined in the profile. So probably zero is fine. Still, do these actually work when the property is missing from the JSON? I'm not familiar with the behavior of the JSON library you use, so I wonder whether trying to reference e.g. `["width"]` will simply return default value (e.g. zero) or will raise an exception or something.